File: ios/RDRCTEventEmitter.m
//
//  RDRCTEventEmitter.m
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 12.07.2023.
//

#import "RDRCTEventEmitter.h"
#import "RDRCTUtils.h"

@interface RDRCTEventEmitter ()
@property(nonatomic, strong) NSMutableArray *pendingEvents;
@end

NSString *const RDRCTPendingEventName = @"com.relateddigital.onPendingEvent";

NSString *const RDRCTEventNameKey = @"name";
NSString *const RDRCTEventBodyKey = @"body";

@implementation RDRCTEventEmitter

static RDRCTEventEmitter *sharedEventEmitter_;

+ (void)load {
  sharedEventEmitter_ = [[RDRCTEventEmitter alloc] init];
}

+ (RDRCTEventEmitter *)shared {
  return sharedEventEmitter_;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    self.pendingEvents = [NSMutableArray array];
  }
  return self;
}

- (void)sendEventWithName:(NSString *)eventName {
  @synchronized(self.pendingEvents) {
    [self.pendingEvents addObject:@{RDRCTEventNameKey : eventName}];
    [self notifyPendingEvents];
  }
}

- (void)sendEventWithName:(NSString *)eventName body:(id)body {
  @synchronized(self.pendingEvents) {
    [self.pendingEvents
        addObject:@{RDRCTEventNameKey : eventName, RDRCTEventBodyKey : body}];
    [self notifyPendingEvents];
  }
}

- (void)notifyPendingEvents {
  [self.bridge enqueueJSCall:@"RCTDeviceEventEmitter"
                      method:@"emit"
                        args:@[ RDRCTPendingEventName ]
                  completion:nil];
}

- (NSArray *)takePendingEventsWithType:(NSString *)type {
  @synchronized(self.pendingEvents) {
    NSMutableArray *events = [NSMutableArray array];
    for (id event in [self.pendingEvents copy]) {
      if ([event[RDRCTEventNameKey] isEqualToString:type]) {
        [events addObject:event[RDRCTEventBodyKey]];
        [self.pendingEvents removeObject:event];
      }
    }
    return events;
  }
}

- (void)onRelatedDigitalListenerAddedForType:(NSString *)type {
  @synchronized(self.pendingEvents) {
    for (id event in [self.pendingEvents copy]) {
      if ([event[RDRCTEventNameKey] isEqualToString:type]) {
        [self notifyPendingEvents];
        break;
      }
    }
  }
}

@end

File: ios/RDRCTEventEmitter.swift
//
//  RDRCTEventEmitter.swift
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 14.07.2023.
//

import Foundation

import React

@objc public class RDRCTEventEmitter: NSObject {

    @objc static let RDRCTPendingEventName = "com.relateddigital.onPendingEvent"
    @objc static let RDRCTEventNameKey = "name"
    @objc static let RDRCTEventBodyKey = "body"
    @objc public weak var bridge: RCTBridge?
    @objc public var pendingEvents: [[String: Any]] = []

    // @objc public let pendingEventsQueue = DispatchQueue(label: "com.relateddigital.pendingEventsQueue", attributes: .concurrent)

    override init() {
        super.init()
        self.pendingEvents = [[String: Any]]()
    }

    @objc public static let shared = RDRCTEventEmitter()

    @objc public func sendEvent(withName eventName: String) {
        // self.pendingEventsQueue.async(flags: .barrier) { [weak self] in
        self.pendingEvents.append([Self.RDRCTEventNameKey: eventName])
        self.notifyPendingEvents()
        // }
    }

    @objc public func sendEvent(withName eventName: String, body: Any?) {
        // self.pendingEventsQueue.async(flags: .barrier) { [weak self] in
        var event: [String: Any] = [Self.RDRCTEventNameKey: eventName]
        if let body = body {
            event[Self.RDRCTEventBodyKey] = body
        }
        self.pendingEvents.append(event)
        self.notifyPendingEvents()
        // }

    }

    @objc public func notifyPendingEvents() {
        self.bridge?.enqueueJSCall(
            "RCTDeviceEventEmitter",
            method: "emit",
            args: [Self.RDRCTPendingEventName],
            completion: nil)
    }

    @objc public func takePendingEvents(withType type: String) -> [Any] {
        var events: [Any] = []
        // pendingEventsQueue.async(flags: .barrier) { [weak self] in
        // guard let self = self else { return }
        self.pendingEvents = self.pendingEvents.filter { event in
            if event[Self.RDRCTEventNameKey] as? String == type {
                if let eventBody = event[Self.RDRCTEventBodyKey] {
                    events.append(eventBody)
                }
                return false
            } else {
                return true
            }
        }
        // }
        return events
    }

    @objc public func onRelatedDigitalListenerAdded(forType type: String) {
        // pendingEventsQueue.async(flags: .barrier) { [weak self] in
        // guard let self = self else { return }
        if self.pendingEvents.contains(where: { $0[Self.RDRCTEventNameKey] as? String == type }) {
            self.notifyPendingEvents()
        }
        // }
    }
}
